### ISSUE TITLE
Decorator friendly syntax.

### ISSUE THREAD

**Issue by proofit404** [proofit404/dependencies#289]
### Problem

If we have a lot of decorators (which is a good OOP pattern), we would end with a lot of nested `Injector` subclasses to build the composition.

Such classes usually have an argument with the same name.

```python
@attrs
class Book:
    title = attrib()
    author = attrib()

    def is_written_by(self, writer):
        return self.author == writer


@attrs
class LoggedBook:
    book = attrib()
    logger = attrib()

    def is_written_by(self, writer):
        self.logger.debug('Arguments: %s', writer)
        result = self.book.is_written_by(writer)
        self.logger.debug('Result: %s', result)
        return result


@attrs
class CachedBook:
    book = attrib()
    cache = {}

    def is_written_by(self, writer):
        if writer not in self.cache:
            self.cache[writer] = self.book.is_written_by(writer)
        return self.cache[writer]
```

```python
class Container(Injector):
    tmp = CachedBook
    book = this.Container.tmp
    
    class Container(Injector):
        tmp = LoggedBook
        book = this.Container.tmp
        logger = Logger
        
        class Container(Injector):
            tmp = Book
            title = 'foo'
            author = 'bar'

book = Container.tmp
```

### Solution

It should be a simpler way to define `Injector` subclass with exactly the same semantic.

```python
from dependencies import Injector, decorate

class Container(Injector):
    book = decorate(CachedBook, LoggedBook, Book)
    logger = Logger
    title = 'foo'
    author = 'bar'
    
    
book = Container.book
```

**Comment by wasinski** [comment]
This is absolutely a brilliant idea!

One impediment I can see though is that it might be that decorators do not match the decoree in the case of the kwarg name even though they do implement the same interface.

Let me give you an example:

```
class CommandHandler(ABC):
    @abstractmethod
    def handle(self, command): ...

class LoggingCommandHandlerDecorator(CommandHandler):
    command_handler: CommandHandler
    logger: Logger

    def handle(self, command):
         self.logger.log(f"executing command {command})
         return self.handler.handle(command)

class CreateUserAccountCommandHandler(CommandHandler):
    user_service: UserService

    def handle(self, command):
         self.user_service.create(name=command.name)
```
So here I would like to be able to decorate any CommandHandler because I have multiple in my app.
Something like:
```
class Container(Injector):
     user_service = UserService
     logger = Logger

     create_user_account = decorate(LoggingCommandHandlerDecorator, CreateUserAccountCommandHandler)
     suspend_user_account = decorate(LoggingCommandHandlerDecorator, SuspendUserAccountCommandHandler)
     [...]
```
So from my understanding in your example, you're matching the kwargs from the `Container.book` to `Cached.book`, in my example, unfortunately, `Container.create_user_account` does not match (by name) `LoggingCommandHandlerDecorator.command_handler). Maybe explicitly stating what is being decorated would solve it? `decorate(LoggingCommandHandlerDecorator, CreateUserAccountCommandHandler, decoree=command_handler)`

One more feature comes to my mind: often DI containers allow to decorate all of the registered classes which implement a given interface with given decorators in a single definition, so in my example one definition would be sufficient to decorate all of the `CommandHandlers`. Maybe this could also be done by the kwarg match?

**Comment by proofit404** [comment]
Thank you for such detailed explanation!

I totally agree that equality between name inside dependency injection scope and name of the decorator argument creates too much restrictions for that feature to be useful.

If we take into consideration that `decorate` should be able to receive arbitrary number of arguments, it became hard to invent a variant of it API. It should reduce boilerplate to define it. And at the same time have almost the same semantics with nested `Injector` and `this` objects.

I'll take my time to think about possible ways to workaround problem you mentioned.

We should not start implementing the feature until we figured it.

Thanks!

**Comment by proofit404** [comment]
```quote
One more feature comes to my mind: often DI containers allow to decorate all of the registered classes which implement a given interface with given decorators in a single definition, so in my example one definition would be sufficient to decorate all of the CommandHandlers. Maybe this could also be done by the kwarg match?
```
First of all, this will add a lot of magic to the object creation. Probably, It'll hard to find the place where this automatic decoration happens in a big project. This approach reminds me about aspect oriented programming. Based on my experience I can't find such feature helping me to improve the clarity of the code base.

If you have different experience in that field, feel free to create a separate ticket so we could discuss it there :wink:

**Comment by proofit404** [comment]
@wasinski Maybe if we would be able to use [shield object](https://github.com/proofit404/dependencies/milestone/10) together with `decorate` that would solve the problem?

```python
@attrs
class Book:
    title = attrib()
    author = attrib()

    def is_written_by(self, writer):
        return self.author == writer


@attrs
class LoggedItem:
    item = attrib() # <== Constructor argument miss match.
    logger = attrib()

    def is_written_by(self, writer):
        self.logger.debug('Arguments: %s', writer)
        result = self.item.is_written_by(writer)
        self.logger.debug('Result: %s', result)
        return result


@attrs
class CachedBook:
    book = attrib()
    cache = {}

    def is_written_by(self, writer):
        if writer not in self.cache:
            self.cache[writer] = self.book.is_written_by(writer)
        return self.cache[writer]
```
So we could define it like this:
```python
from dependencies import Injector, decorate, shield, this

class Container(Injector):
    book = decorate(CachedBook, shield(LoggedItem, item=this.book), Book)
    logger = Logger
    title = 'foo'
    author = 'bar'
    
book = Container.book
```
But this creates an unclear situation **at what scope `this` object points?**

**Comment by proofit404** [comment]
Well, this will contradict with another rule of shield object. So it's not an option.

https://github.com/proofit404/dependencies/issues/441

**Comment by proofit404** [comment]
The `decorate` object looks way to similar to the [`pluck` object](https://github.com/proofit404/dependencies/issues/309).

```python
from dependencies import Injector, decorate

class Container(Injector):
    book = decorate(CachedBook, LoggedBook, Book)
    logger = Logger
    title = 'foo'
    author = 'bar'
``` 
The code above looks way to similar to 
```python
from dependencies import Injector, pluck

class Container(Injector):
    book = pluck(CachedBook, book=pluck(LoggedBook, book=Book))
    logger = Logger
    title = 'foo'
    author = 'bar'
```

**Comment by proofit404** [comment]
Superseded by #481
